tools/run_cibuildwheel.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python
+
+import logging
+import os
+import runpy
+from shlex import quote
+import sys
+
+_LOG = logging.getLogger(__name__)
+
+
+def set_env(name: str, value: str) -> None:
+    _LOG.info("set %s = %s", name, value)
+    os.environ[name] = value
+
+
+def append_inner_env(name: str, value: str) -> None:
+    existing = os.environ.get("CIBW_ENVIRONMENT", "")
+    set_env("CIBW_ENVIRONMENT", f"{existing} {name}={value}")
+
+
+def setup_env() -> None:
+    # Per https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html,
+    # we need gcc 5.1 for stable C++11 ABI.
+    # manylinux1 has gcc 4.8.2.
+    # manylinux2010 and manylinux2014 have a backported gcc at version
+    # 8+. Their glibcs are old; they achieve support for newer gcc by
+    # statically linking backported newer glibc functions.
+    # manylinux_2_24 has a "plain" install of gcc 6.x.
+
+    set_env("CIBW_MANYLINUX_X86_64_IMAGE", "manylinux2010")
+    set_env("CIBW_MANYLINUX_I686_IMAGE", "manylinux2010")
+
+    # manylinux2014 is the oldest available for these archs.
+    set_env("CIBW_MANYLINUX_AARCH64_IMAGE", "manylinux2014")
+    set_env("CIBW_MANYLINUX_PPC64LE_IMAGE", "manylinux2014")
+    set_env("CIBW_MANYLINUX_S390X_IMAGE", "manylinux2014")
+
+    # This would normally be in root files like setup.cfg or setup.py, but all
+    # our files are in bindings/python so cibuildwheel can't figure it out on
+    # its own
+    set_env("CIBW_PROJECT_REQUIRES_PYTHON", ">=3.6")
+    # Per https://cibuildwheel.readthedocs.io/en/stable/cpp_standards/, 10.9
+    # is required for full C++11 support
+    set_env("MACOSX_DEPLOYMENT_TARGET", "10.9")
+
+    # Set up tests.
+    # NB: cibuildwheel expands {project} to the project root.
+    set_env(
+        "CIBW_TEST_COMMAND",
+        'bash -c "cd \'{project}/bindings/python\' && python test.py"',
+    )
+
+    # Set up boost.
+
+    # TODO: support pre-existing boost in $BOOST_ROOT, for mac and windows.
+
+    boost_version = os.environ["BOOST_VERSION"]
+
+    if sys.platform == "win32":
+        boost_root = "c:/boost"
+    else:
+        boost_root = "/tmp/boost"
+    boost_build_path = f"{boost_root}/tools/build"
+
+    before_all_commands = [
+        f"./tools/setup_boost.sh {quote(boost_version)} {quote(boost_root)}",
```

Comment:
do we download boost every time this is run? even if we've run it before?

Reply:
On linux, `cibuildwheel` spawns a docker container and does the builds there. On other platforms it runs build commands nakedly.

I thought I'd written that in comments, but I guess it got lost in refactors.